tests/unit/test_charm.py

```diff
@@ -168,71 +175,120 @@ def test_given_when_bessd_pebble_ready_then_expected_pebble_plan_is_applied(  #
         self.assertEqual(expected_plan, updated_plan)
 
     @patch(f"{MULTUS_LIBRARY_PATH}.KubernetesMultusCharmLib.is_ready")
-    @patch("ops.model.Container.exec")
     def test_given_can_connect_to_bessd_when_bessd_pebble_ready_then_ip_route_is_created(
-        self, patch_exec, patch_is_ready
+        self, patch_is_ready
     ):
+        call_list = []
+
+        def bessd_cmds_handler(args: testing.ExecArgs) -> testing.ExecResult:
+            call_list.append(
+                HandleExecCallArgs(
+                    command=args.command,
+                    timeout=args.timeout,
+                    environment=args.environment,
+                )
+            )
+            return testing.ExecResult(exit_code=0)
+
+        self.harness.handle_exec("bessd", [], handler=bessd_cmds_handler)
```

Comment:
@benhoyt is this the recommended approach to use `handle_exec`?

Reply:
Not exactly. This looks like a direct port of the existing patch code, which is fine. But I'd recommend using the command matching part of `handle_exec`. For example:

```python
    @patch(f"{MULTUS_LIBRARY_PATH}.KubernetesMultusCharmLib.is_ready")
    def test_given_can_connect_to_bessd_when_bessd_pebble_ready_then_ip_route_is_created(
        self, patch_is_ready
    ):
        ip_command = ["ip", "route", "replace", "default", "via", CORE_GW_IP, "metric", "110"]
        self.harness.handle_exec("bessd", ip_command, result=0)
        self.harness.handle_exec("bessd", ["iptables-legacy"], result=0)
        self.harness.handle_exec("bessd", ["/opt/bess/bessctl/bessctl"], result=0)

        patch_is_ready.return_value = True

        self.harness.container_pebble_ready(container_name="bessd")
```

Or if you really want to ensure that the `ip route ...` command was called at least once:

```python
    @patch(f"{MULTUS_LIBRARY_PATH}.KubernetesMultusCharmLib.is_ready")
    def test_given_can_connect_to_bessd_when_bessd_pebble_ready_then_ip_route_is_created(
        self, patch_is_ready
    ):
        ip_route_called = False

        def ip_handler(args: testing.ExecArgs) -> testing.ExecResult:
            nonlocal ip_route_called
            ip_route_called = True
            return testing.ExecResult(exit_code=0)

        ip_command = ["ip", "route", "replace", "default", "via", CORE_GW_IP, "metric", "110"]
        self.harness.handle_exec("bessd", ip_command, handler=ip_handler)
        self.harness.handle_exec("bessd", ["iptables-legacy"], result=0)
        self.harness.handle_exec("bessd", ["/opt/bess/bessctl/bessctl"], result=0)

        patch_is_ready.return_value = True

        self.harness.container_pebble_ready(container_name="bessd")

        self.assertTrue(ip_route_called)
```

That second example is not all that far from what you have in the PR now (you're testing for a side effect), but it's better to use a more specific command prefix in the `handle_exec` call.

Reply:
Thanks for the precious hints. I updated the PR, now it should be better.